### PR TITLE
Add keyboard/input-source

### PR DESCRIPTION
**Pull request by jpnurmi**
This replaces the `setxkbmap` call that was coupled with the keyboard layout setting, to let the Desktop Installer GUI control the desktop's input source so that the user can test layouts before applying the selection.

### CODE REVIEW

**Comment by jpnurmi** [comment]
With this, we can have the Desktop Installer in sync with GNOME Control Center.

Screenshot from 2022-08-15 18-51-59

File: subiquity/server/controllers/tests/test_desktop.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        self.assertEqual(self.controller.model.username, "baz")
+
+    @patch("subiquity.server.controllers.desktop.arun_command")
+    async def test_set_input_source(self, mock_arun_command):
+        self.app.opts.dry_run = False
+        self.controller.model.username = "bar"
+        await self.controller.set_input_source("fr", "latin9")
+        gsettings = [
+            "gsettings", "set", "org.gnome.desktop.input-sources", "sources",
+            "[('xkb','fr+latin9')]"
```
Comment (dbungert) [inline subiquity/server/controllers/tests/test_desktop.py:47]: A check on a version without a variant would be nice.
Some parts of the subiquity code use `parameterized` to build a list of input data, such as https://github.com/canonical/subiquity/blob/d6cbbda68503e3c1642288f292269f4763d50e78/subiquitycore/tests/test_async_helpers.py#L29

**Review by dbungert — CHANGES_REQUESTED** [review]
JP, I appreciate the direct Subiquity work.
I have some feedback.
I'm not intrinsically opposed to a Desktop controller, but in this case I'm not sure we want it.  Feel free to discuss further, either here or privately.

File: subiquity/models/tests/test_desktop.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        self.model = DesktopModel(self.tmp_dir())
+        self.assertEqual(self.model.username, 'ubuntu')
+
+    def testCasperConf(self):
+        tmpdir = self.tmp_dir()
+        conf = self.tmp_path('etc/casper.conf', tmpdir)
+        os.makedirs(os.path.dirname(conf))
+        with open(conf, 'w') as fp:
+            fp.write('''
+export USERNAME="someone"
```
[the remark this answers is not in the document]
Comment (jpnurmi) [reply subiquity/models/tests/test_desktop.py:34]: Thanks, great suggestion. A username argument sounds much better than this. :)

File: subiquity/server/controllers/desktop.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+from subiquitycore.utils import arun_command
+
+from subiquity.common.apidef import API
+from subiquity.common.types import KeyboardSetting
+from subiquity.server.controller import SubiquityController
+
+log = logging.getLogger("subiquity.server.controllers.desktop")
+
+run_as_command = """
```
[the remark this answers is not in the document]
Comment (jpnurmi) [reply subiquity/server/controllers/desktop.py:29]: It's certainly not fun to debug when things go wrong. Anyway, consider this temporary until Python 3.9+ is available.

**Review by jpnurmi — COMMENTED** [review]
Thanks for the feedback. The idea is to extend this with auto-login settings and add support for flavors. KDE, MATE, Cinnamon, XFCE all have different ways to store these settings.

File: subiquity/server/controllers/tests/test_desktop.py (hunk 1 above)
Comment (jpnurmi) [reply subiquity/server/controllers/tests/test_desktop.py:47]: the combination of `@parameterized.expand()` and `@patch()` doesn't seem to work with async functions. a simple test case to illustrate the problem:

```py
import os
import unittest

from parameterized import parameterized
from unittest.mock import patch


class TestParameterizedMock(unittest.IsolatedAsyncioTestCase):

    @parameterized.expand([(0,)])
    @patch("os.seteuid")
    def test_parameterized_mock(self, uid, mock_seteuid):
        os.seteuid(uid)  # OK: calls the mock
        mock_seteuid.assert_called_once_with(uid)

    @parameterized.expand([(0,)])
    @patch("os.seteuid")
    async def test_parameterized_mock_async(self, uid, mock_seteuid):
        os.seteuid(uid)  # NOT OK: PermissionError: [Errno 1] Operation not permitted
        mock_seteuid.assert_called_once_with(uid)


if __name__ == '__main__':
    unittest.main()
```

**Comment by jpnurmi** [comment]
### INPUT-SOURCE

- GNOME: `gsettings org.gnome.desktop.input-sources sources`
- MATE: `gsettings org.mate.peripherals-keyboard-xkb.kbd layouts`
- CINNAMON: `gsettings org.gnome.libgnomekbd.keyboard layouts`
- XFCE: `$XDG_CONFIG_HOME/xfce4/xfconf/xfce-perchannel-xml/keyboard-layout.xml`
- KDE: `$XDG_DATA_HOME/kded5/keyboard/session/layout_memory.xml`

### COLOR-SCHEME

- GNOME: `gsettings org.gnome.desktop.interface color-scheme`
- (XDG: `xdg-desktop-portal` via D-Bus?)

### AUTO-LOGIN

- GDM: `/etc/gdm3/custom.conf`:
- KDM: `/etc/kde4/kdm/kdmrc`
- LXDM: `/etc/lxdm/lxdm.conf`
- LIGHTDM: `/etc/lightdm`
- SDDM: `/etc/sddm.conf`

File: subiquity/server/controllers/tests/test_desktop.py (hunk 1 above)
Comment (dbungert) [reply subiquity/server/controllers/tests/test_desktop.py:47]:
```
@parameterized.expand([(0,)])
async def test_parameterized_mock_async(self, uid):
    with patch("os.seteuid") as mock_seteuid:
        os.seteuid(uid)  # OK
        mock_seteuid.assert_called_once_with(uid)
```
`with patch` is a nice construct and would help here.

File: subiquity/server/controllers/desktop.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        """ Save the current state of the model so it can be loaded later.
+        Currently this function is called automatically by .configured().
+        """
+        return self.model.username
+
+    def deserialize(self, username: str) -> None:
+        """ Loads the last-known state of the model. """
+        self.model.username = username
+
+    async def input_source_POST(self, data: KeyboardSetting) -> None:
```
[the remark this answers is not in the document]
Comment (jpnurmi) [reply subiquity/server/controllers/desktop.py]: I envisioned that there would be more settings and that we could have the run-as hack hidden here, but we could move this to the keyboard controller instead.

File: subiquity/server/controllers/desktop.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/server/controllers/desktop.py]: > @ogayot - I know a version of this is in the codebase already, but I'm not wild about this python script as string in the code. Is there another viable structure, say with `multiprocessing` or as an external script that we ship with the snap?

Yes, for sure we can move the code to a `.py` script (under bin/ or scripts/) and call it. I guess we would want to polish the script if we want to make it a `.py` though.

I don't have experience with `multiprocessing` but if there is a good alternative, it's all good to me.

[the remark this answers is not in the document]
Comment (jpnurmi) [reply subiquity/server/controllers/desktop.py]: I got it working with systemd-run.

**Comment by jpnurmi** [comment]
```bash
ubuntu@ubuntu:~$ gsettings get org.gnome.desktop.input-sources sources
[('xkb', 'us')]
ubuntu@ubuntu:~$ curl --unix-socket /run/subiquity/socket -X POST a/keyboard/input_source?user=%22ubuntu%22 -d '{"layout":"fi","variant":"mac"}'
ubuntu@ubuntu:~$ gsettings get org.gnome.desktop.input-sources sources
[('xkb', 'fi+mac')]
```

```
2022-08-18 17:08:16,591 DEBUG root:37 start: subiquity/Keyboard/input_source_POST: 
2022-08-18 17:08:16,591 DEBUG subiquitycore.utils:92 arun_command called: ['systemd-run', '--wait', '--uid=999', '--setenv=DISPLAY=:0', '--setenv=XDG_RUNTIME_DIR=/run/user/999', '--setenv=DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/999/bus', 'gsettings', 'set', 'org.gnome.desktop.input-sources', 'sources', "[('xkb','fi+mac')]"]
2022-08-18 17:08:16,605 DEBUG subiquitycore.utils:101 arun_command ['systemd-run', '--wait', '--uid=999', '--setenv=DISPLAY=:0', '--setenv=XDG_RUNTIME_DIR=/run/user/999', '--setenv=DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/999/bus', 'gsettings', 'set', 'org.gnome.desktop.input-sources', 'sources', "[('xkb','fi+mac')]"] exited with code 0
2022-08-18 17:08:16,606 DEBUG root:37 finish: subiquity/Keyboard/input_source_POST: SUCCESS: 200 null
2022-08-18 17:08:16,606 INFO aiohttp.access:233  [18/Aug/2022:17:08:16 +0000] "POST /keyboard/input_source?user=%22ubuntu%22 HTTP/1.1" 200 189 "-" "curl/7.84.0"
```

**Comment by jpnurmi** [comment]
@dbungert All comments have been addressed. Is it starting to look better?

**Comment by mwhudson** [comment]
This looks pretty good (apart from the summary being wrong now :wink:), I guess the only question I have is why is this a separate API call rather than something that is done every time the keyboard layout is set?

**Comment by jpnurmi** [comment]
I'm proposing to keep it separate because applying an explicit keyboard layout setting would lose automatic keyboard layout lookup from language (https://github.com/canonical/subiquity/pull/1368). The GUI has to adjust the system's input source when arriving at the keyboard layout screen, before the user has accepted the proposed layout, to make sure that the test input field works as expected.

**Review by dbungert — APPROVED** [review]

**Comment by mwhudson** [comment]
> The GUI has to adjust the system's input source when arriving at the keyboard layout screen, before the user has accepted the proposed layout, to make sure that the test input field works as expected.

Ah right

**Pull request merged**